python/bandpassUtils.py

```diff
@@ -0,0 +1,251 @@
+    # precompute coefficients
+    b = np.asmatrix([[k**i for i in order_range] for k in range(-half_window, half_window+1)])
+    m = np.linalg.pinv(b).A[deriv] * rate**deriv * np.math.factorial(deriv)
+    # pad the signal at the extremes with
+    # values taken from the signal itself
+    firstvals = y[0] - np.abs(y[1:half_window+1][::-1] - y[0] )
+    lastvals = y[-1] + np.abs(y[-half_window-1:-1][::-1] - y[-1])
+    y = np.concatenate((firstvals, y, lastvals))
+    return np.convolve(m[::-1], y, mode='valid')
+
+def buildLens(lensDir, addLosses=True):
+    """
+    Build each individual lens. Assumes there are files l*Glass.dat, l*BBAR.dat, and a subdirectory with loss files.
+    Returns a bandpass object.
+    """
+    lens = Bandpass()
+    # Read the glass base file.
+    glassfile = glob(os.path.join(lensDir, 'l*Glass.dat'))
+    if len(glassfile) != 1:
+        raise ValueError('Expected a single glass file in this directory, found: ', glassfile)
+    glassfile = glassfile[0]
+    glass = Bandpass()
+    glass.readThroughput(glassfile)
+    # Smooth the glass response.
+    smoothGlassSb = savitzky_golay(glass.sb, 31, 3)
+    lens.setBandpass(glass.wavelen, smoothGlassSb)
+    # Read the broad band antireflective (BBAR) coatings files.
+    bbarfiles = glob(os.path.join(lensDir, 'l*BBAR.dat'))
+    bbars = Bandpass()
+    bbars.readThroughputList(bbarfiles)
+    # Multiply the bbars by the glass.
+    wavelen, sb = lens.multiplyThroughputs(bbars.wavelen, bbars.sb)
+    lens.setBandpass(wavelen, sb)
+    # Add losses.
+    if addLosses:
+        lossfiles = glob(os.path.join(lensDir, 'l*Losses", ''*.dat'))
+        loss = Bandpass()
+        loss.readThroughputList(lossfiles)
+        wavelen, sb = lens.multiplyThroughputs(loss.wavelen, loss.sb)
+        lens.setBandpass(wavelen, sb)
+    return lens
+
+def buildMirror(mirrorDir, addLosses=True):
+    """
+    Build a mirror throughput curve. Assumes there are (a) *Losses.dat file and a *Ideal.dat file.
+    Returns a bandpass object.
+    """
+    # Read the mirror reflectance curve.
+    mirrorfile = glob(os.path.join(mirrorDir, 'm*Ideal.dat'))
+    if len(mirrorfile) != 1:
+        raise ValueError('Expected a single mirror file in directory %s, found: ' %mirrorDir, mirrorfile)
+    mirrorfile = mirrorfile[0]
+    mirror = Bandpass()
+    mirror.readThroughput(mirrorfile)
+    if addLosses:
+        lossfiles = glob(os.path.join(mirrorDir, 'm*Losses.dat'))
+        loss = Bandpass()
+        loss.readThroughputList(lossfiles)
+        wavelen, sb = mirror.multiplyThroughputs(loss.wavelen, loss.sb)
+        mirror.setBandpass(wavelen, sb)
+    return mirror
+
+def buildAtmosphere(atmosDir):
+    """
+    Read the atmosphere throughput curve.
+    Returns a bandpass object.
+    """
+    atmofile = os.path.join(atmosDir, 'pachonModtranAtm_12.dat')
+    atmo = Bandpass()
+    atmo.readThroughput(atmofile)
+    return atmo
+
+def buildHardwareAndSystem(defaultDirs, addLosses=True):
```

Comment:
docstring should specify what defaultDirs is (a dict pointing to the directories holding the data files associated with each throughput component)